crates/wasmtime/src/component/types.rs

```diff
@@ -0,0 +1,1025 @@
+use crate::component::func::{self, Lift, Memory, Options};
+use crate::component::values::{
+    self, Enum, Expected, Flags, List, Option, Record, Tuple, Union, Val, Variant,
+};
+use crate::store::StoreOpaque;
+use crate::ValRaw;
+use anyhow::{anyhow, bail, Context, Error, Result};
+use std::collections::HashMap;
+use std::fmt;
+use std::iter;
+use std::ops::Deref;
+use std::sync::Arc;
+use wasmtime_component_util::{DiscriminantSize, FlagsSize};
+use wasmtime_environ::component::{
+    ComponentTypes, InterfaceType, TypeEnumIndex, TypeExpectedIndex, TypeFlagsIndex,
+    TypeInterfaceIndex, TypeRecordIndex, TypeTupleIndex, TypeUnionIndex, TypeVariantIndex,
+};
+
+#[derive(Clone)]
+pub struct Handle<T> {
+    index: T,
+    types: Arc<ComponentTypes>,
+}
+
+impl<T: fmt::Debug> fmt::Debug for Handle<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Handle")
+            .field("index", &self.index)
+            .finish()
+    }
+}
+
+impl<T: PartialEq> PartialEq for Handle<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.index == other.index && Arc::ptr_eq(&self.types, &other.types)
```

Comment:
I think that this may lead to unfortunately surprising results in that a string type is only considered equivalent if it comes from the same `types` origin but it's actually unique across all components. I think that this is sufficient for a "succeed fast" test but this may need to be expanded for deep equality checking as well.

Reply:
I'm not sure I understand this.  `Type::String` doesn't have a `Handle<T>`, so this code wouldn't be reached in that case.  Am I misunderstanding what you meant by "string type"?

Reply:
Ah sorry bad example, but instead I think this would consider `(u32, u32)` not an equivalent type if it comes from two different components.

Reply:
I've address this for tuple, option, and expected.  Not sure if I should do the same for e.g. record, though.  See my comment in the other conversation about type equality.

Reply:
I think that the tuple/option/expected bits are probably not going to be necessary in the long-term if the original issue here is fixed. I think this is fine to leave as a FIXME with a follow-up issue for now though.

Otherwise though the current logic in tuple/option/expected I would expect to be baked in here or otherwise into some trait impl on `T` perhaps. I don't think we'd want to duplicate things and have some using `PartialEq` and some using manual impls.